File: sdk/go/provider/tempmailpro.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const tempmailproBase = "https://tempmailpro.us/api/v1"

type tempmailproMailboxResponse struct {
	Data struct {
		Token     string `json:"token"`
		Address   string `json:"address"`
		ExpiresAt any    `json:"expires_at"`
		CreatedAt any    `json:"created_at"`
	} `json:"data"`
}

type tempmailproListResponse struct {
	Data []map[string]any `json:"data"`
}

type tempmailproDetailResponse struct {
	Data map[string]any `json:"data"`
}

func tempmailproJSON(method, u string, body []byte, out any) error {
	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	req, err := http.NewRequest(method, u, reader)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("tempmailpro: http %d", resp.StatusCode)
	}
	if out == nil {
		return nil
	}
	return json.Unmarshal(raw, out)
}

func tempmailproString(v any) string {
	if v == nil {
		return ""
	}
	return strings.TrimSpace(fmt.Sprint(v))
}

func tempmailproFlatten(raw map[string]any, recipient string) map[string]any {
	id := raw["id"]
	if id == nil || tempmailproString(id) == "" {
		id = raw["message_id"]
	}
	from := raw["from_address"]
	if from == nil || tempmailproString(from) == "" {
		from = raw["from_name"]
	}
	return map[string]any{
		"id":          id,
		"from":        from,
		"to":          recipient,
		"subject":     raw["subject"],
		"text":        raw["body_text"],
		"html":        raw["body_html"],
		"date":        raw["received_at"],
		"attachments": raw["attachments"],
	}
}

func TempmailproGenerate() (*CreatedMailbox, error) {
	var data tempmailproMailboxResponse
	if err := tempmailproJSON(http.MethodPost, tempmailproBase+"/mailbox/create", []byte("{}"), &data); err != nil {
		return nil, err
	}
	email := strings.TrimSpace(data.Data.Address)
	token := strings.TrimSpace(data.Data.Token)
	if email == "" || !strings.Contains(email, "@") || token == "" {
		return nil, fmt.Errorf("tempmailpro: invalid mailbox response")
	}
	return &CreatedMailbox{
		Channel:   "tempmailpro",
		Email:     email,
		Token:     token,
		ExpiresAt: data.Data.ExpiresAt,
		CreatedAt: tempmailproString(data.Data.CreatedAt),
	}, nil
}

func tempmailproFetchDetail(token, messageID string) (map[string]any, error) {
	u := fmt.Sprintf("%s/mailbox/%s/emails/%s", tempmailproBase, url.PathEscape(token), url.PathEscape(messageID))
	var data tempmailproDetailResponse
	if err := tempmailproJSON(http.MethodGet, u, nil, &data); err != nil {
		return nil, err
	}
	if data.Data == nil {
		return nil, fmt.Errorf("tempmailpro: empty message detail")
	}
	return data.Data, nil
}

func TempmailproGetEmails(token, email string) ([]NormEmail, error) {
	mailboxToken := strings.TrimSpace(token)
	address := strings.TrimSpace(email)
	if mailboxToken == "" {
		return nil, fmt.Errorf("tempmailpro: empty token")
	}
	if address == "" {
		return nil, fmt.Errorf("tempmailpro: empty email")
	}

	u := fmt.Sprintf("%s/mailbox/%s/emails", tempmailproBase, url.PathEscape(mailboxToken))
	var data tempmailproListResponse
	if err := tempmailproJSON(http.MethodGet, u, nil, &data); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(data.Data))
	for _, row := range data.Data {
		raw := row
		if id := tempmailproString(row["id"]); id != "" {
			if detail, err := tempmailproFetchDetail(mailboxToken, id); err == nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(tempmailproFlatten(raw, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/tempmailpro.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "tempmailpro";
const API_BASE = "https://tempmailpro.us/api/v1";

interface MailboxResponse {
  data?: {
    token?: string;
    address?: string;
    expires_at?: number;
    created_at?: number;
  };
}

interface ListResponse {
  data?: Array<Record<string, unknown>>;
}

interface DetailResponse {
  data?: Record<string, unknown>;
}

async function fetchJSON<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      Accept: "application/json",
      ...(init?.headers || {}),
    },
  });
  if (!response.ok) {
    throw new Error(`tempmailpro http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

function flattenMessage(
  raw: Record<string, unknown>,
  recipient: string,
): Record<string, unknown> {
  return {
    id: raw.id || raw.message_id,
    from: raw.from_address || raw.from_name,
    to: recipient,
    subject: raw.subject,
    text: raw.body_text,
    html: raw.body_html,
    date: raw.received_at,
    attachments: raw.attachments,
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const data = await fetchJSON<MailboxResponse>(`${API_BASE}/mailbox/create`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({}),
  });
  const box = data.data || {};
  const email = String(box.address || "").trim();
  const token = String(box.token || "").trim();
  if (!email || !email.includes("@") || !token) {
    throw new Error("tempmailpro: invalid mailbox response");
  }
  return {
    channel: CHANNEL,
    email,
    token,
    expiresAt: box.expires_at,
    createdAt: box.created_at == null ? undefined : String(box.created_at),
  };
}

async function fetchDetail(
  token: string,
  messageId: string,
): Promise<Record<string, unknown> | null> {
  try {
    const data = await fetchJSON<DetailResponse>(
      `${API_BASE}/mailbox/${encodeURIComponent(token)}/emails/${encodeURIComponent(messageId)}`,
    );
    return data.data && typeof data.data === "object" ? data.data : null;
  } catch {
    return null;
  }
}

export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const mailboxToken = String(token || "").trim();
  const address = String(email || "").trim();
  if (!mailboxToken) throw new Error("tempmailpro: empty token");
  if (!address) throw new Error("tempmailpro: empty email");

  const data = await fetchJSON<ListResponse>(
    `${API_BASE}/mailbox/${encodeURIComponent(mailboxToken)}/emails`,
  );
  const rows = Array.isArray(data.data) ? data.data : [];
  const emails: Email[] = [];
  for (const row of rows) {
    const id = row.id == null ? "" : String(row.id).trim();
    const detail = id ? await fetchDetail(mailboxToken, id) : null;
    emails.push(
      normalizeEmail(flattenMessage(detail || row, address), address),
    );
  }
  return emails;
}
